xarray/coding/calendar_ops.py

```diff
@@ -0,0 +1,263 @@
+            name=dim,
+        )
+        # Remove duplicate timestamps, happens when reducing the number of days
+        out = out.isel({dim: np.unique(out[dim], return_index=True)[1]})
+    elif align_on == "date":
+        new_times = convert_times(
+            time.data,
+            get_date_type(calendar, use_cftime=use_cftime),
+            raise_on_invalid=False,
+        )
+        out[dim] = new_times
+
+        # Remove NaN that where put on invalid dates in target calendar
+        out = out.where(out[dim].notnull(), drop=True)
+
+    if missing is not None:
+        time_target = date_range_like(time, calendar=calendar, use_cftime=use_cftime)
+        out = out.reindex({dim: time_target}, fill_value=missing)
+
+    # Copy attrs but remove `calendar` if still present.
+    out[dim].attrs.update(time.attrs)
+    out[dim].attrs.pop("calendar", None)
+    return out
+
+
+def _datetime_to_decimal_year(times, calendar=None):
+    """Convert a datetime DataArray to decimal years according to its calendar or the given one.
+
+    Decimal years are the number of years since 0001-01-01 00:00:00 AD.
+    Ex: '2000-03-01 12:00' is 2000.1653 in a standard calendar, 2000.16301 in a "noleap" or 2000.16806 in a "360_day".
+    """
+    from ..core.dataarray import DataArray
+
+    calendar = calendar or times.dt.calendar
+
+    if is_np_datetime_like(times.dtype):
+        times = times.copy(data=convert_times(times.values, get_date_type("standard")))
+
+    def _make_index(time):
+        year = int(time.dt.year[0])
+        doys = cftime.date2num(time, f"days since {year:04d}-01-01", calendar=calendar)
+        return DataArray(
+            year + doys / _days_in_year(year, calendar),
+            dims=time.dims,
+            coords=time.coords,
+            name="time",
+        )
+
+    return times.groupby("time.year").map(_make_index)
+
+
+def interp_calendar(source, target, dim="time"):
+    """Interpolates a DataArray/Dataset to another calendar based on decimal year measure.
+
+    Each timestamp in source and target are first converted to their decimal year equivalent
+    then source is interpolated on the target coordinate. The decimal year is the number of
+    years since 0001-01-01 AD.
+    Ex: '2000-03-01 12:00' is 2000.1653 in a standard calendar or 2000.16301 in a 'noleap' calendar.
+
+    This method should only be used when the time (HH:MM:SS) information of time coordinate is not important.
+
+    Parameters
+    ----------
+    source: Union[DataArray, Dataset]
+      The source data to interpolate, must have a time coordinate of a valid dtype (:py:class:`numpy.datetime64`  or :py:class:`cftime.datetime` objects)
+    target: DataArray or DatetimeIndex or CFTimeIndex
+      The target time coordinate of a valid dtype (np.datetime64 or cftime objects)
+    dim : str
+      The time coordinate name.
+
+    Return
+    ------
+    Union[DataArray, Dataset]
+      The source interpolated on the decimal years of target,
```

Comment:
Right! I added a test in both function  : if there is a 0 in `time.dt.year` and the target calendar is in ["gregorian", "proleptic_gregorian", "julian", "standard"], a ValueError is raised.

Technically, with cftime 1.5 we could create a `gregorian` calendar with a year 0, but the current code only passes the calendar name as a string, so this information is lost... To be done later, upon request.

Reply:
Awesome, thanks!

> Technically, with cftime 1.5 we could create a gregorian calendar with a year 0, but the current code only passes the calendar name as a string, so this information is lost... To be done later, upon request.

That's true.  My current opinion on the `has_year_zero` option in xarray is that we officially only support the options dictated by the CF conventions.  While it may be possible to manually construct `cftime.datetime` objects that follow other conventions for year zero, these would be complicated to deal with fully in xarray, not least because it would require defining our own serialization conventions (we'd need to encode not just the calendar type but also the year zero convention).